no-extra-parens confilct with no-confusing-arrow

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

ESLint - v2.9.0 (and 2.10.1 - see below)  
parser - default  
this code:

``` javascript
          set(val) {
            let result = val;
            if (_.isArray(val)) {
              result = _.map(val, (str) => /*THERE*/(_.isString(str) ? str.replace(/\(|\)|\-|\s/g, '') : str));
            }
            this.setDataValue(fieldName, JSON.stringify(result));
          }
```

 -- give "no-extra-parens" error.

and this:

``` javascript
          set(val) {
            let result = val;
            if (_.isArray(val)) {
              result = _.map(val, /*THERE*/(str) => _.isString(str) ? str.replace(/\(|\)|\-|\s/g, '') : str);
            }
            this.setDataValue(fieldName, JSON.stringify(result));
          }
```

 -- give "no-confusing-arrow" error

``` json
    "no-confusing-arrow": [
      2,
      {
        "allowParens": true
      }
    ],
    "no-extra-parens": [
      2,
      "all",
      {
        "conditionalAssign": true,
        "nestedBinaryExpressions": false
      }
    ],
```

 -- part of configuration

I already have tried update ESlint to 2.10.1, and change configuration to:  

```
    "no-extra-parens": [
      2,
      "all",
      {
        "conditionalAssign": false,
        "nestedBinaryExpressions": false,
        "returnAssign": false
      }
    ],
```

It did not help.  
I expect, that "no extra parens" error would not appear in that case.

I'm also experiencing this problem:

**What version of ESLint are you using?**
v3.0.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
/* eslint-env es6 */
/* eslint no-confusing-arrow: [ "error", { allowParens: true } ], "no-extra-parens": [ "error", "all", { conditionalAssign : false, returnAssign : false, nestedBinaryExpressions : false } ] */
```

**What did you do? Please include the actual source code causing the issue.**
Consider the following examples:

```js
/* eslint-env es6 */
/* eslint no-confusing-arrow: [ "error", { allowParens: true } ], "no-extra-parens": [ "error", "all", { conditionalAssign : false, returnAssign : false, nestedBinaryExpressions : false } ] */

// this correctly warns about `no-extra-parens`
var a1 = ( 1 );

// this correctly warns about `no-confusing-arrow`
var b1 = x => x ? 1 : 2;

// Expected: this should be valid (as per http://eslint.org/docs/rules/no-confusing-arrow#options)
// Actual: it warns about `no-extra-parens`
var c1 = x => ( x ? 1 : 2 );

/* eslint-disable no-extra-parens */
// this now works fine, no warnings
var c2 = x => ( x ? 1 : 2 );
/* eslint-enable no-extra-parens */
```

**What did you expect to happen?**
The above test file should output **2** errors, for lines 5 & 8.
`var c1 = x => ( x ? 1 : 2 );` should be valid with both `no-confusing-arrow` & `no-extra-parens` enabled.

**What actually happened? Please include the actual, raw output from ESLint.**
`var c1 = x => ( x ? 1 : 2 );` triggered the `no-extra-parens` warning message.

```
~/dev/test-cases/eslint-conflict-6196/test.js
   5:10  error  Gratuitous parentheses around expression                       no-extra-parens
   8:10  error  Arrow function used ambiguously with a conditional expression  no-confusing-arrow
  12:15  error  Gratuitous parentheses around expression                       no-extra-parens

✖ 3 problems (3 errors, 0 warnings)
```

--- a/docs/rules/no-extra-parens.md
+++ b/docs/rules/no-extra-parens.md
@@ -22,6 +22,7 @@ This rule has an object option for exceptions to the `"all"` option:
 * `"returnAssign": false` allows extra parentheses around assignments in `return` statements
 * `"nestedBinaryExpressions": false` allows extra parentheses in nested binary expressions
 * `"ignoreJSX": "none|all|multi-line|single-line"` allows extra parentheses around no/all/multi-line/single-line JSX components. Defaults to `none`.
+* `"enforceForArrowConditionals": false` allows extra parentheses around ternary expressions which are the body of an arrow function
 
 ### all
 
@@ -163,6 +164,17 @@ Examples of **correct** code for this rule with the `all` and `{ "ignoreJSX": "s
 /* eslint no-extra-parens: ["error", "all", { ignoreJSX: "single-line" }] */
 const Component = (<div />)
 const Component = (<div><p /></div>)
+```
+
+### enforceForArrowConditionals
+
+Examples of **correct** code for this rule with the `"all"` and `{ "enforceForArrowConditionals": false }` options:
+
+```js
+/* eslint no-extra-parens: ["error", "all", { "enforceForArrowConditionals": false }] */
+
+const b = a => 1 ? 2 : 3;
+const d = c => (1 ? 2 : 3);
 ```
 
 ### functions

--- a/lib/rules/no-extra-parens.js
+++ b/lib/rules/no-extra-parens.js
@@ -44,7 +44,8 @@ module.exports = {
                                 conditionalAssign: { type: "boolean" },
                                 nestedBinaryExpressions: { type: "boolean" },
                                 returnAssign: { type: "boolean" },
-                                ignoreJSX: { enum: ["none", "all", "single-line", "multi-line"] }
+                                ignoreJSX: { enum: ["none", "all", "single-line", "multi-line"] },
+                                enforceForArrowConditionals: { type: "boolean" }
                             },
                             additionalProperties: false
                         }
@@ -67,6 +68,9 @@ module.exports = {
         const NESTED_BINARY = ALL_NODES && context.options[1] && context.options[1].nestedBinaryExpressions === false;
         const EXCEPT_RETURN_ASSIGN = ALL_NODES && context.options[1] && context.options[1].returnAssign === false;
         const IGNORE_JSX = ALL_NODES && context.options[1] && context.options[1].ignoreJSX;
+        const IGNORE_ARROW_CONDITIONALS = ALL_NODES && context.options[1] &&
+            context.options[1].enforceForArrowConditionals === false;
+
         const PRECEDENCE_OF_ASSIGNMENT_EXPR = precedence({ type: "AssignmentExpression" });
         const PRECEDENCE_OF_UPDATE_EXPR = precedence({ type: "UpdateExpression" });
 
@@ -428,6 +432,13 @@ module.exports = {
 
             ArrowFunctionExpression(node) {
                 if (isReturnAssignException(node)) {
+                    return;
+                }
+
+                if (node.body.type === "ConditionalExpression" &&
+                    IGNORE_ARROW_CONDITIONALS &&
+                    !isParenthesisedTwice(node.body)
+                ) {
                     return;
                 }
 

--- a/tests/lib/rules/no-extra-parens.js
+++ b/tests/lib/rules/no-extra-parens.js
@@ -369,6 +369,10 @@ ruleTester.run("no-extra-parens", rule, {
             "/>)"
         ].join("\n"), options: ["all", { ignoreJSX: "multi-line" }] },
 
+        // ["all", { enforceForArrowConditionals: false }]
+        { code: "var a = b => 1 ? 2 : 3", options: ["all", { enforceForArrowConditionals: false }], parserOptions: { ecmaVersion: 6 } },
+        { code: "var a = (b) => (1 ? 2 : 3)", options: ["all", { enforceForArrowConditionals: false }], parserOptions: { ecmaVersion: 6 } },
+
         {
             code: "let a = [ ...b ]",
             parserOptions: { ecmaVersion: 2015 }
@@ -921,6 +925,32 @@ ruleTester.run("no-extra-parens", rule, {
         ].join("\n"), "const Component = <div>\n<p />\n</div>", "JSXElement", 1, {
             options: ["all", { ignoreJSX: "none" }]
         }),
+
+        // ["all", { enforceForArrowConditionals: true }]
+        {
+            code: "var a = (b) => (1 ? 2 : 3)",
+            parserOptions: { ecmaVersion: 6 },
+            options: ["all", { enforceForArrowConditionals: true }],
+            errors: [
+                {
+                    message: "Gratuitous parentheses around expression."
+                }
+            ],
+            output: "var a = (b) => 1 ? 2 : 3"
+        },
+
+        // ["all", { enforceForArrowConditionals: false }]
+        {
+            code: "var a = (b) => ((1 ? 2 : 3))",
+            parserOptions: { ecmaVersion: 6 },
+            options: ["all", { enforceForArrowConditionals: false }],
+            errors: [
+                {
+                    message: "Gratuitous parentheses around expression."
+                }
+            ],
+            output: "var a = (b) => (1 ? 2 : 3)"
+        },
 
         // https://github.com/eslint/eslint/issues/8175
         invalid(

